Fix work-around item non-sync issue on AMD

zoranzhao opened the pull request:
Summary: Otherwise it will break FSDP code paths

Test Plan:
unit test

see next diff for print message
```
sh ./scripts/lufang/amd/small_repro.sh
ROCM_GET_SCALAR_ITEM_SYNC=1 sh ./scripts/lufang/amd/small_repro.sh
```

It will log "====== Async mode ======" or "====== Sync mode ======" correspondingly

Differential Revision: D60995134

houseroad reviewed (approved):
Looks good. Thanks for fixing FSDP path.

zoranzhao committed:
Fix work-around item non-sync issue on AMD (#133054)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/133054

Otherwise it will break FSDP code paths

Test Plan:
unit test

see next diff for print message
```
sh ./scripts/lufang/amd/small_repro.sh
HIP_DOUBLE_SYNC_ON_LOCAL_SCALE_DENSE=1 sh ./scripts/lufang/amd/small_repro.sh
```

It will log "====== Async mode ======" or "====== Sync mode ======" correspondingly

Reviewed By: houseroad

Differential Revision: D60995134

zoranzhao commented on aten/src/ATen/native/cuda/CUDAScalar.cu:
```diff
@@ -11,36 +11,48 @@
-#else
-  // TODO(lufang): Tensor.item() on AMD HIP is not synced in the Recsys models.
-  // This is just a short term workaround. Issue is tracked as FBA-388 on the AMD side.
-  auto cpu_self = self.cpu();
```
Mapped to a barrier and invoke a cache flash here
